plan/plan.go

```diff
@@ -52,36 +52,47 @@ type Changes struct {
 // processing. It returns a copy of Plan with the changes populated.
 func (p *Plan) Calculate() *Plan {
 	changes := &Changes{}
-
-	// Ensure all desired records exist. For each desired record make sure it's
-	// either created or updated.
-	for _, desired := range p.Desired {
-		// Get the matching current record if it exists.
-		current, exists := recordExists(desired, p.Current)
-
-		// If there's no current record create desired record.
-		if !exists {
-			changes.Create = append(changes.Create, desired)
-			continue
+	usedDesired := map[int]bool{}
+	usedCurrent := map[int]bool{}
+
+	for c, current := range p.Current {
+		for d, desired := range p.Desired {
+			if current.DNSName == desired.DNSName && current.Target == desired.Target &&
+				current.SuitableType() == desired.SuitableType() {
+				usedDesired[d] = true
+				usedCurrent[c] = true
+				log.Debugf("Skipping endpoint %v because target has not changed", desired)
+			}
 		}
+	}
 
-		// If there already is a record update it if it changed.
-		if desired.Target != current.Target {
-			changes.UpdateOld = append(changes.UpdateOld, current)
-
-			desired.RecordType = current.RecordType // inherit the type from the dns provider
-			desired.MergeLabels(current.Labels)     // inherit the labels from the dns provider, including Owner ID
-			changes.UpdateNew = append(changes.UpdateNew, desired)
+	for c, current := range p.Current {
+		if _, found := usedCurrent[c]; found {
 			continue
 		}
-
-		log.Debugf("Skipping endpoint %v because target has not changed", desired)
+		for d, desired := range p.Desired {
```

Comment:
I still don't understand how empty-string record type should be handled so that suitableType() won't be used. For example, are "foo->bar/CNAME" conflicts with "foo->baz/". If the later can become "foo->baz/ALIAS", does it mean that plan computation need to know what ALIAS is? IMO the first thing that need to have some standardization of types and do not use empty type at all. For example, let the provider convert standard CNAME to provider-specific ALIAS back and forth. Or maybe there is already an order there but I don't get it. Once we agree on that, the solution I can see is to extend policies with ability to validate/pre-filter records before plan computation (i.e. Apply(changes) -> ApplyBefore(current, desired) + ApplyAfter(changes)) and have a policy, which detects this case and either leaves the first CNAME or deletes all conflicting CNAME records and prints warning. Maybe also let the providers install their own policies to handle vendor-specific rules